### PR TITLE
copy/multiple: instanceCopyCopy honor `UpdateCompressionAlgorithms`

### PR DESCRIPTION
**Pull request by flouthoc**
Needed By: https://github.com/containers/skopeo/pull/2047

* Users can set compression formats while performing `copy` in such cases
`instanceCopyCopy` must honor UpdateCompressionAlgorithms.

* copy/single: set requiredCompression if compressionFormat is changed
 - Modifies `copy/single` to correctly use the feature from
https://github.com/containers/image/pull/2023, that is if compression is
changed blob must be resued only if it matches required compression.

* oci_index: init Annotations if nil and required
 - If `UpdateCompressionAlgorithms` is set then `Annotations` map must be
initialized if its not set.

Integration tests for `e2e` feature added here: https://github.com/containers/skopeo/pull/2047#issue-1809369600

### CODE REVIEW

**Comment by flouthoc** [comment]
@mtrmac PTAL

**Review by mtrmac — COMMENTED** [review]
Consider a simple `skopeo copy --format oci` of an existing image: In that case we should preserve the original blobs (or reuse blobs at destination, if possible), and add the OCI annotations.

So while `copy --dest-compress-format zstd` is the most salient use case, I think this should work by collecting compression algorithms from all copies.

- IIRC the `BlobInfo` returned from `copyLayer` always contains an algorithm, or `Unknown`/`Uncompressed.
- Then, `copyLayers` can deduplicate the values into a set; and return that value.
- `copySingleImage` can return the relevant algorithms (optionally add a `copySingleImageResult struct` so that we don’t have that many unnamed return values)
- and that can then be used for `ListEdit`

**Comment by mtrmac** [comment]
> Consider a simple `skopeo copy --format oci` of an existing image: In that case we should preserve the original blobs (or reuse blobs at destination, if possible), and add the OCI annotations.

(If the source is not already OCI, it is probably v2s1 or v2s2, which only support gzip, so that _seems_ like a non-issue. But the copy can reuse zstd blobs at the destination, creating an OCI image that is mixed gzip + Zstd, or possibly Zstd-only.)

**Review by mtrmac — COMMENTED** [review]
Also I think the shortcut in `OptimizeDestinationImageAlreadyExists` must not be taken if `require…Match`.

**Comment by mtrmac** [comment]
Ugh, the `OptimizeDestinationImageAlreadyExists` path really is a pain… When we do a full copy, we do get algorithm data from all blobs as a side-effect. When we skip that, all of that is missing and needs to be obtained elsewhere. Options I can see:
1. Teach `internal/image.SourcedImage` (and `manifest` implementations) to return a list of algorithms from a manifest. That’s quite possible (we have `compressionMIMETypeSet` tables already), probably cleanest, but it’s another new API addition, and the data we would get that way comes from a different source than from ordinary copies.
2. Take the `list.Instance().ReadOnly.CompressionAlgorithms` data, which was recently computed, pass it from the outside down to `copySingleImage`, and have `copySingleImage` return that to the caller on the `Optimize…` path. Technically that would work, except that the value would not be available in the non-multi-image callers of `copySingleImage`, i.e. that would introduce an inconsistency in behavior.
3. Take the conceptually-misplaced heuristic determining `CompressionAlgorithm` at the start of `copyLayer`; extract it to a separate helper function; and call that in `compareImageDestinationManifestEqual` to build the data. That way the algorithm values we return is determined exactly the same way on the `Optimize…` path and on the other one, and always available.

I think 3. is the cleanest way to do this right now; and if we ever replace the “conceptually-misplaced heuristic” (which serves as a nice example that technical debt never goes away on its own) with a proper `manifest` API addition, that would naturally handle both code paths.

File: internal/manifest/oci_index.go
Comment (flouthoc) [reply internal/manifest/oci_index.go:149]: moving this to `addCompressionAnnotations` would require the function to accept `Manifest` directly cause sending the value of `index.Manifests[targetIndex].Annotations` will not work if its `nil`

Comment (mtrmac) [reply internal/manifest/oci_index.go:149]: I thought sending a `*map…` would work — but I didn’t test that.

**Comment by flouthoc** [comment]
@mtrmac I have not updated tests yet but this is what I am thinking. WDYT ?

File: copy/multiple.go
Comment (mtrmac) [inline copy/multiple.go:51]: All of this about sending `require…Match` down to `copySingleImage` is relevant for the larger work, but not necessary for this PR, unless I’m missing something.

(And per #1987 , I default to thinking that it should be set only on the clone copies, but I’m not sure about that. … and there’s also the option for this PR to just build the infrastructure in `copySingleImage`, always setting it to `false`, and for #1987 to actually add the logic that sets it.)

---

If you intentionally want to make this PR add the two things (setting “update algorithms” correctly, and adding the `require…Match` infrastructure), that’s also a choice that can be made, just let me know whether you prefer this to do one or both.

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:51]: I think this can be built on `internal/set.Set` instead of manually.

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:66]: (In the PR that adds the new parameter) it might be worthwhile to add a `copySingleImageOptions struct`
, primarily so that the boolean is named. OTOH it’s just one boolean right now, and `copyBlob…` is much worse, so blocking on that wouldn’t be fair.

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:192]: Stylistically I’d prefer this to be one of the checks inside this top-level `if`, so that the “checking if we can skip copying” line is logged (and that should log the new value as well).

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:215]: (Non-blocking: I’m tempted to say `copyLayers` should do this.)

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:270]: I’d probably add it to `wipResult` in one place, either after the loop, or at the very end when forming the final value, to reduce duplication / risk of drift.

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:410]: Non-blocking: I’d prefer a `return nil, …` on error paths, so that there is no data dependency and no ambiguity about what we return.

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:405]: Absolutely non-blocking: `res` might be too short a name for such a long function`. `compressionAlgos`? Or maybe define the value much later in the function, so that it has a short scope.

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:455]: This runs in a concurrently-executing goroutine, so it is invalid to modify a shared `res` variable from here. Instead, pass it through a new field in `copyLayerData`, and collect (+deduplicate?) it when processing the goroutine results.

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:625]: I think this can be a stand-alone function, not a method; that would make it more convenient to potentially unit-test.

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:627]: If we have that data, we should return it, not ignore it. Or clearly document that the caller needs to handle that case, and actually do that in the new caller.

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:653]: Please move this comment to the new function.

**Review by mtrmac — COMMENTED** [review]
A quick pass, not a careful review — so far all the primary design points look correct.

File: copy/multiple.go
Comment (flouthoc) [reply copy/multiple.go:51]: I think `require..Match` is needed for this case as well, consider the case where registry/infocache already contains blobs in a different compression and I am doing `--dest-format-compress` so afaik skopeo skips pushing blobs with `blob already exists` when `require..match` is not implemented (I assume it is reusing `gzip` blobs candidates). I am able to see this problem.

#### Without `require..match` does not copies new compression
```console
[fl@fedora bin]$ ./skopeo copy --all --format oci --dest-compress-format zstd --src-tls-verify=false --dest-tls-verify=false docker://localhost:5000/list docker://localhost:5000/list
Getting image list signatures
Copying 2 of 2 images in list
Copying image sha256:da39e651fdc155fbd53152dee387b91d4fd1e76da3137ccbd332f98e6c552a63 (1/2)
Getting image source signatures
Copying blob 8a117f3d2e73 skipped: already exists  
Copying blob 5119035169e2 skipped: already exists  
Copying config 58dce7cc03 done   | 
Writing manifest to image destination
Copying image sha256:f87412510ffdd07383a3b433ab58159522a4ffa3048f3557f55708ec90c8008c (2/2)
Getting image source signatures
Copying blob 33bbc77ac1e2 skipped: already exists  
Copying blob 5119035169e2 skipped: already exists  
Copying config 0044dbde8e done   | 
Writing manifest to image destination
Writing manifest list to image destination
Storing list signatures
``` 

#### With `require..match` correctly works
```console
[fl@fedora bin]$ ./skopeo copy --all --format oci --dest-compress-format zstd --src-tls-verify=false --dest-tls-verify=false docker://localhost:5000/list docker://localhost:5000/list
Getting image list signatures
Copying 2 of 2 images in list
Copying image sha256:da39e651fdc155fbd53152dee387b91d4fd1e76da3137ccbd332f98e6c552a63 (1/2)
Getting image source signatures
Copying blob 8a117f3d2e73 done   | 
Copying blob 5119035169e2 done   | 
Copying config 58dce7cc03 done   | 
Writing manifest to image destination
Copying image sha256:f87412510ffdd07383a3b433ab58159522a4ffa3048f3557f55708ec90c8008c (2/2)
Getting image source signatures
Copying blob 33bbc77ac1e2 done   | 
Copying blob 5119035169e2 done   | 
Copying config 0044dbde8e done   | 
Writing manifest to image destination
Writing manifest list to image destination
Storing list signatures
```

Comment (flouthoc) [reply copy/multiple.go:51]: I just added it cause it was constantly causing issues for me while testing locally, I suspect others might also hit this issue while trying to do the same, would it be an issue if we add it here and expose the option on frontend later in a different PR ? I think exposing part would be much simpler then.

File: copy/single.go
Comment (mtrmac) [reply copy/single.go:51]: … and probably inlined in the two loops that collect the data.

File: copy/multiple.go
Comment (mtrmac) [reply copy/multiple.go:51]: For the record, repeating from other discussion:
- The underlying `require…Match` functionality is definitely necessary for #1987 .
- I don’t, at this point, think that we should change the current behavior for `--dest-compress-format` (that’s https://github.com/containers/buildah/issues/4613 ); it should be a new option. I don’t feel _very_ strongly about this, and it might be worth discussing, but it also seems unrelated to the primary feature we are working towards at this very moment.

File: copy/single.go
Comment (mtrmac) [reply copy/single.go:627]: Looking at more details, only manifest’s `.LayerInfos()` never sets this field; only `pkg/blobcache` sets it in `LayerInfosForCopy`.

So the new caller in `compareImageDestinationManifestEqual` can be build assuming `==nil` above; and in that case this function can a simpler `compressionAlgorithmFromMIMEType`, leaving the `if` only in the `copyLayer` caller.

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:153]: (Absolutely non-blocking: if it is unconditional like this, the field can be set directly set in the `imageCopier{` constructor above.)

**Review by mtrmac — COMMENTED** [review]
A full review now — ~nothing new.

File: copy/single.go
Comment (flouthoc) [reply copy/single.go]: `Set.Add(` ensures that no duplicates are there ( I have added a test to verify that ) however `compression.Algorithm` is not a `comparable` hence I have created a `Set` using `compression.Algorithm.Name()` and its working fine.

Later I'm using `AlgorithmsByNames` to retrieve a slice of  `Algorithms`.

File: copy/multiple.go
Comment (flouthoc) [reply copy/multiple.go]: Cool I have just implemented infra for now and passing `false` as default value, will implement this in a different PR.

File: copy/single.go
Comment (flouthoc) [reply copy/single.go:53]: > OTOH it’s just one boolean right now, and copyBlob… is much worse, so blocking on that wouldn’t be fair.

Cool I'll implement this later in a different PR.

File: internal/manifest/oci_index.go
Comment (flouthoc) [reply internal/manifest/oci_index.go]: done.

File: internal/set/set_test.go
Comment (mtrmac) [inline internal/set/set_test.go:41]: Non-blocking: I think this can be tested in `TestAdd` or `TestValues` (or both). OTOH this works fine, and the test file is small enough.

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:184]: Please log this value in the debug log above, so that the decision can be explained.

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:394]: Absolutely non-blocking: This can be set only just before it’s computed (around `diffIDs := `)

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:629]: The part of the comment above this line should stay in `copyLayer`; it is not relevant to the “image already exists” case.

File: pkg/compression/compression.go
Comment (mtrmac) [inline pkg/compression/compression.go:56]: Non-blocking:

At this point I think this can be an internal helper in `copy/single.go`. Adding more lookup functions to `pkg/compression` is a bit unattractive:

Context: Importing `pkg/compression` adds all compression implementations, including `c/storage/pkg/chunked/compression`, as dependencies, and it’s unclear to me that that’s always a good idea. Compare how `transports` contains a registry, and `transports/alltransports` adds the dependencies: the caller can choose which one to use. (OTOH the argument that each consumer of `pkg/compression` should always support all algorithms might be stronger, I don’t have a firm opinion.)

---

An alternative could be to modify `ListEdit.*CompressionAlgorithms` to only contain names — that would also get rid of the “algorithm name is known but we don’t have an implementation” failure paths — but I think we don’t need to tinker with that now. In practice `compressionAlgorithmFromMIMEType`, or the data from `blob/compression.go`, are always supported algorithms, so that error path is not worth worrying about or making larger changes now.

**Review by mtrmac — COMMENTED** [review]
Thanks!

Last comments on a debug log and a misplaced comment.

File: internal/set/set_test.go
Comment (flouthoc) [reply internal/set/set_test.go]: done.

File: copy/single.go
Comment (mtrmac) [inline copy/single.go:495]: Oops, I apologize for missing this earlier … `cld.destInfo` is only valid if `cld.err` == nil`; so the new code should be added below.

**Review by mtrmac — COMMENTED** [review]
LGTM otherwise, feel free to merge without another round.

File: copy/single.go
Comment (flouthoc) [reply copy/single.go:492]: done.

**Comment by flouthoc** [comment]
I don't have merge access, @containers/image-maintainers PTAL

**Review by mtrmac — APPROVED** [review]
Thanks!

**Pull request merged**
